### ISSUE TITLE
Delete API: Rows in instance_label_map label, mount, service_event, instance_link, external_handler_external_handler_process_map tables are never marked as removed

### ISSUE THREAD

**Issue by LLParse** [rancher/rancher#6659]
**Rancher Version:** master @ 4e3bc12ec8fd66bd910efbaffacd2a43cb408637

**Docker Version:** 1.12.3

**OS and where are the hosts located? (cloud, bare metal, etc):** ubuntu 14.04

**Setup Details: (single node rancher vs. HA rancher, internal DB vs. external DB)** single node

**Environment Type: (Cattle/Kubernetes/Swarm/Mesos)** cattle

**Steps to Reproduce:**
1. Launch a stack
2. Delete a stack

**Results:** 
The instance_label_map label, mount, service_event, instance_link, external_handler_external_handler_process_map rows aren't marked as removed, preventing deletion of instance rows. Database cleanup logic doesn't delete these rows and must skip other removed rows being referenced by foreign key constraints.

**Expected:**
The rows are marked as removed and the database cleanup logic deletes the rows after the configured cutoff time.

**Comment by LLParse** [comment]
Related to https://github.com/rancher/rancher/issues/6276

**Comment by deniseschannon** [comment]
@sangeethah When testing this issue, please verify that the rancher db server housekeeping issue (#5129) is addressed.

**Comment by cjellick** [comment]
1.`mount` get removed when a volume is purged. No change needed.
2. `instance_link` gets removed when instance is purged. No change need
3. `instance_label_map` - will just delete in the cleanup code
4. `label` - will just delete in the cleanup thread
5. `service_event` - will just delete in the cleanup thread
6. `external_handler_external_handler_process_map` - not sure yet

**Comment by sangeethah** [comment]
Tested with v1.2.0-pre4-rc15

I am still not able to see entries from instance,service,environment being cleaned up.
There is Foreign key errors seen when attempting to deleting entries from instances and healthcheck_instance.

Deleted services/Environment entries seem to be in "removed" state and never transition on "purged" state. Will these entries be picked up cleanup thread for deletion?



cattle-error.log:2016-11-26 16:26:05,266 WARN  [:] [] [] [] [TaskScheduler-1] [i.c.p.core.cleanup.TableCleanup     ] [Rows Skipped] volume=1 service_index=100 service=100 environment=100 agent=96 account=96 

```
cattle-debug.log:2016-11-26 16:25:44,905 INFO  [:] [] [] [] [TaskScheduler-1] [i.c.p.core.cleanup.TableCleanup     ] [Rows Deleted] instance_label_map=100 label=53 
cattle-debug.log:2016-11-26 16:25:44,928 INFO  [:] [] [] [] [TaskScheduler-1] [i.c.p.core.cleanup.TableCleanup     ] [Rows Deleted] external_handler_external_handler_process_map=15 
cattle-debug.log:2016-11-26 16:25:44,969 INFO  [:] [] [] [] [TaskScheduler-1] [i.c.p.core.cleanup.TableCleanup     ] Cleanup process_instance tables [cutoff=Fri Nov 25 16:25:44 UTC 2016] 
cattle-debug.log:2016-11-26 16:25:44,998 INFO  [:] [] [] [] [TaskScheduler-1] [i.c.p.core.cleanup.TableCleanup     ] Cleanup event tables [cutoff=Sat Nov 12 16:25:44 UTC 2016] 
cattle-debug.log:2016-11-26 16:25:45,025 INFO  [:] [] [] [] [TaskScheduler-1] [i.c.p.core.cleanup.TableCleanup     ] Cleanup audit_log tables [cutoff=Sun Aug 28 16:25:44 UTC 2016] 
cattle-debug.log:2016-11-26 16:25:45,178 INFO  [:] [] [] [] [TaskScheduler-1] [i.c.p.core.cleanup.TableCleanup     ] Cleanup service_log tables [cutoff=Fri Nov 25 16:25:44 UTC 2016] 
cattle-debug.log:2016-11-26 16:26:01,768 ERROR [:] [] [] [] [TaskScheduler-1] [i.c.p.core.cleanup.TableCleanup     ] SQL [delete from `healthcheck_instance` where `healthcheck_instance`.`id` in (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)]; Cannot delete or update a parent row: a foreign key constraint fails (`cattle`.`service_event`, CONSTRAINT `fk_service_event__healthcheck_instance_id` FOREIGN KEY (`healthcheck_instance_id`) REFERENCES `healthcheck_instance` (`id`) ON DELETE NO ACTION ON UPDATE ) 
cattle-debug.log:2016-11-26 16:26:04,835 ERROR [:] [] [] [] [TaskScheduler-1] [i.c.p.core.cleanup.TableCleanup     ] SQL [delete from `instance` where `instance`.`id` in (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)]; Cannot delete or update a parent row: a foreign key constraint fails (`cattle`.`instance_label_map`, CONSTRAINT `fk_instance_label_map__instance_id` FOREIGN KEY (`instance_id`) REFERENCES `instance` (`id`) ON DELETE NO ACTION ON UPDATE NO ACTION) 
cattle-debug.log:2016-11-26 16:26:05,266 INFO  [:] [] [] [] [TaskScheduler-1] [i.c.p.core.cleanup.TableCleanup     ] Cleanup other tables [cutoff=Sat Nov 26 16:15:44 UTC 2016] 
cattle-debug.log:2016-11-26 16:26:05,266 INFO  [:] [] [] [] [TaskScheduler-1] [i.c.p.core.cleanup.TableCleanup     ] [Rows Deleted] certificate=5 config_item_status=1249 dynamic_schema=33 external_handler=3 healthcheck_instance_host_map=319 instance_host_map=1804 instance_link=46 ip_address_nic_map=1783 mount=1149 nic=1804 port=3250 service_consume_map=179 service_expose_map=1774 service_log=24450 volume_storage_pool_map=2820 ip_address=1783 volume=2819 deployment_unit=1218 service=39 credential=97  
cattle-debug.log:2016-11-26 16:26:05,266 WARN  [:] [] [] [] [TaskScheduler-1] [i.c.p.core.cleanup.TableCleanup     ] [Rows Skipped] volume=1 service_index=100 service=100 environment=100 agent=96 account=96  
cattle-error.log:2016-11-26 16:26:01,768 ERROR [:] [] [] [] [TaskScheduler-1] [i.c.p.core.cleanup.TableCleanup     ] SQL [delete from `healthcheck_instance` where `healthcheck_instance`.`id` in (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)]; Cannot delete or update a parent row: a foreign key constraint fails (`cattle`.`service_event`, CONSTRAINT `fk_service_event__healthcheck_instance_id` FOREIGN KEY (`healthcheck_instance_id`) REFERENCES `healthcheck_instance` (`id`) ON DELETE NO ACTION ON UPDATE ) 
cattle-error.log:2016-11-26 16:26:04,835 ERROR [:] [] [] [] [TaskScheduler-1] [i.c.p.core.cleanup.TableCleanup     ] SQL [delete from `instance` where `instance`.`id` in (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)]; Cannot delete or update a parent row: a foreign key constraint fails (`cattle`.`instance_label_map`, CONSTRAINT `fk_instance_label_map__instance_id` FOREIGN KEY (`instance_id`) REFERENCES `instance` (`id`) ON DELETE NO ACTION ON UPDATE NO ACTION) 
cattle-error.log:2016-11-26 16:26:05,266 WARN  [:] [] [] [] [TaskScheduler-1] [i.c.p.core.cleanup.TableCleanup     ] [Rows Skipped] volume=1 service_index=100 service=100 environment=100 agent=96 account=96  
```

**Comment by cjellick** [comment]
yes @sangeethah services and envs will get picked up

**Comment by sangeethah** [comment]
Tested with latest build from master:

There is no FK constraints when Table cleanup is done. I see entries from service/environment and instance table being cleaned up.

But there are few entries that are  not getting cleaned up from instance/service/environment. Some of the instances are Lb instances but i also see regular instances not being removed.

```
[Rows Skipped] healthcheck_instance=6 instance=16 service_index=14 service=10 environment=10 agent=7 account=7  
```
server logs:
```
2016-11-27 20:55:23,270 INFO  [:] [] [] [] [TaskScheduler-1] [i.c.p.core.cleanup.TableCleanup     ] Cleanup other tables [cutoff=Sun Nov 27 20:45:16 UTC 2016] 
2016-11-27 20:55:23,271 INFO  [:] [] [] [] [TaskScheduler-1] [i.c.p.core.cleanup.TableCleanup     ] [Rows Deleted] config_item_status=314 dynamic_schema=42 healthcheck_instance_host_map=12 instance_host_map=347 instance_link=1 ip_address_nic_map=331 mount=187 nic=339 port=479 service_consume_map=122 service_expose_map=388 service_log=2030 volume_storage_pool_map=591 healthcheck_instance=1 ip_address=305 volume=566 deployment_unit=359 instance=304 service_index=243 service=189 environment=69 credential=7  
2016-11-27 20:55:23,272 WARN  [:] [] [] [] [TaskScheduler-1] [i.c.p.core.cleanup.TableCleanup     ] [Rows Skipped] nic=59 healthcheck_instance=5 ip_address=18 volume=27 instance=96 service_index=99 service=49 environment=29 agent=6 account=6  
2016-11-27 21:55:16,425 INFO  [:] [] [] [] [TaskScheduler-1] [i.c.p.core.cleanup.TableCleanup     ] [Rows Deleted] instance_label_map=15 label=4 
2016-11-27 21:55:16,433 INFO  [:] [] [] [] [TaskScheduler-1] [i.c.p.core.cleanup.TableCleanup     ] [Rows Deleted] external_handler_external_handler_process_map=24 
2016-11-27 21:55:16,435 INFO  [:] [] [] [] [TaskScheduler-1] [i.c.p.core.cleanup.TableCleanup     ] Cleanup process_instance tables [cutoff=Sat Nov 26 21:55:16 UTC 2016] 
2016-11-27 21:55:16,437 INFO  [:] [] [] [] [TaskScheduler-1] [i.c.p.core.cleanup.TableCleanup     ] Cleanup event tables [cutoff=Sun Nov 13 21:55:16 UTC 2016] 
2016-11-27 21:55:16,438 INFO  [:] [] [] [] [TaskScheduler-1] [i.c.p.core.cleanup.TableCleanup     ] Cleanup audit_log tables [cutoff=Mon Aug 29 21:55:16 UTC 2016] 
2016-11-27 21:55:16,439 INFO  [:] [] [] [] [TaskScheduler-1] [i.c.p.core.cleanup.TableCleanup     ] Cleanup service_log tables [cutoff=Sat Nov 26 21:55:16 UTC 2016] 
2016-11-27 21:55:16,548 DEBUG [:] [] [] [] [TaskScheduler-1] [i.c.p.core.cleanup.TableCleanup     ] Skipped "cattle"."healthcheck_instance" where id in [1, 5, 6, 7, 8, 9] 
2016-11-27 21:55:16,633 DEBUG [:] [] [] [] [TaskScheduler-1] [i.c.p.core.cleanup.TableCleanup     ] Skipped "cattle"."instance" where id in [285, 338, 348, 367, 380, 382, 388, 431, 432, 433, 1, 430, 472, 473, 474, 475] 
2016-11-27 21:55:16,648 DEBUG [:] [] [] [] [TaskScheduler-1] [i.c.p.core.cleanup.TableCleanup     ] Skipped "cattle"."service_index" where id in [252, 260, 278, 291, 292, 299, 341, 342, 343, 344, 383, 386, 384, 385] 
2016-11-27 21:55:16,661 DEBUG [:] [] [] [] [TaskScheduler-1] [i.c.p.core.cleanup.TableCleanup     ] Skipped "cattle"."service" where id in [180, 188, 214, 218, 221, 224, 240, 241, 253, 254] 
2016-11-27 21:55:16,669 DEBUG [:] [] [] [] [TaskScheduler-1] [i.c.p.core.cleanup.TableCleanup     ] Skipped "cattle"."environment" where id in [72, 74, 85, 86, 87, 88, 100, 101, 111, 112] 
2016-11-27 21:55:16,685 DEBUG [:] [] [] [] [TaskScheduler-1] [i.c.p.core.cleanup.TableCleanup     ] Skipped "cattle"."agent" where id in [5, 14, 15, 16, 17, 18, 19] 
2016-11-27 21:55:16,706 DEBUG [:] [] [] [] [TaskScheduler-1] [i.c.p.core.cleanup.TableCleanup     ] Skipped "cattle"."account" where id in [11, 20, 21, 22, 23, 24, 25] 
2016-11-27 21:55:16,706 INFO  [:] [] [] [] [TaskScheduler-1] [i.c.p.core.cleanup.TableCleanup     ] Cleanup other tables [cutoff=Sun Nov 27 21:45:16 UTC 2016] 
2016-11-27 21:55:16,706 INFO  [:] [] [] [] [TaskScheduler-1] [i.c.p.core.cleanup.TableCleanup     ] [Rows Deleted] config_item_status=15 healthcheck_instance_host_map=8 instance_host_map=85 ip_address_nic_map=92 mount=11 nic=92 port=51 service_expose_map=29 service_log=491 volume_storage_pool_map=38 external_handler=5 ip_address=118 volume=63 deployment_unit=1 instance=112 service_index=108 service=47 environment=28 credential=1  
2016-11-27 21:55:16,706 WARN  [:] [] [] [] [TaskScheduler-1] [i.c.p.core.cleanup.TableCleanup     ] [Rows Skipped] healthcheck_instance=6 instance=16 service_index=14 service=10 environment=10 agent=7 account=7  
2016-11-27 22:55:16,422 INFO  [:] [] [] [] [TaskScheduler-2] [i.c.p.core.cleanup.TableCleanup     ] Cleanup process_instance tables [cutoff=Sat Nov 26 22:55:16 UTC 2016] 
2016-11-27 22:55:16,423 INFO  [:] [] [] [] [TaskScheduler-2] [i.c.p.core.cleanup.TableCleanup     ] Cleanup event tables [cutoff=Sun Nov 13 22:55:16 UTC 2016] 
2016-11-27 22:55:16,424 INFO  [:] [] [] [] [TaskScheduler-2] [i.c.p.core.cleanup.TableCleanup     ] Cleanup audit_log tables [cutoff=Mon Aug 29 22:55:16 UTC 2016] 
2016-11-27 22:55:16,425 INFO  [:] [] [] [] [TaskScheduler-2] [i.c.p.core.cleanup.TableCleanup     ] Cleanup service_log tables [cutoff=Sat Nov 26 22:55:16 UTC 2016] 
2016-11-27 22:55:16,487 DEBUG [:] [] [] [] [TaskScheduler-2] [i.c.p.core.cleanup.TableCleanup     ] Skipped "cattle"."healthcheck_instance" where id in [1, 5, 6, 7, 8, 9] 
2016-11-27 22:55:16,496 DEBUG [:] [] [] [] [TaskScheduler-2] [i.c.p.core.cleanup.TableCleanup     ] Skipped "cattle"."instance" where id in [1, 285, 338, 348, 367, 380, 382, 388, 430, 431, 432, 433, 472, 473, 474, 475] 
2016-11-27 22:55:16,502 DEBUG [:] [] [] [] [TaskScheduler-2] [i.c.p.core.cleanup.TableCleanup     ] Skipped "cattle"."service_index" where id in [252, 260, 278, 291, 292, 299, 341, 342, 343, 344, 383, 384, 385, 386] 
2016-11-27 22:55:16,514 DEBUG [:] [] [] [] [TaskScheduler-2] [i.c.p.core.cleanup.TableCleanup     ] Skipped "cattle"."service" where id in [180, 188, 214, 218, 221, 224, 240, 241, 253, 254] 
2016-11-27 22:55:16,517 DEBUG [:] [] [] [] [TaskScheduler-2] [i.c.p.core.cleanup.TableCleanup     ] Skipped "cattle"."environment" where id in [72, 74, 85, 86, 87, 88, 100, 101, 111, 112] 
2016-11-27 22:55:16,520 DEBUG [:] [] [] [] [TaskScheduler-2] [i.c.p.core.cleanup.TableCleanup     ] Skipped "cattle"."agent" where id in [5, 14, 15, 16, 17, 18, 19] 
2016-11-27 22:55:16,543 DEBUG [:] [] [] [] [TaskScheduler-2] [i.c.p.core.cleanup.TableCleanup     ] Skipped "cattle"."account" where id in [11, 20, 21, 22, 23, 24, 25] 
2016-11-27 22:55:16,543 INFO  [:] [] [] [] [TaskScheduler-2] [i.c.p.core.cleanup.TableCleanup     ] Cleanup other tables [cutoff=Sun Nov 27 22:45:16 UTC 2016] 
2016-11-27 22:55:16,543 INFO  [:] [] [] [] [TaskScheduler-2] [i.c.p.core.cleanup.TableCleanup     ] [Rows Deleted] service_log=508  
2016-11-27 22:55:16,543 WARN  [:] [] [] [] [TaskScheduler-2] [i.c.p.core.cleanup.TableCleanup     ] [Rows Skipped] healthcheck_instance=6 instance=16 service_index=14 service=10 environment=10 agent=7 account=7  
```

```
mysql> select id,name,state,created,removed from environment where state="removed";
+-----+-------------------+---------+---------------------+---------------------+
| id  | name              | state   | created             | removed             |
+-----+-------------------+---------+---------------------+---------------------+
|  72 | test951525rancher | removed | 2016-11-27 20:27:45 | 2016-11-27 20:28:33 |
|  74 | test724411rancher | removed | 2016-11-27 20:28:54 | 2016-11-27 20:30:02 |
|  85 | test471745        | removed | 2016-11-27 20:31:36 | 2016-11-27 20:32:49 |
|  86 | test471745rancher | removed | 2016-11-27 20:31:54 | 2016-11-27 20:32:49 |
|  87 | test127932        | removed | 2016-11-27 20:32:49 | 2016-11-27 20:33:21 |
|  88 | test127932rancher | removed | 2016-11-27 20:33:10 | 2016-11-27 20:33:21 |
| 100 | test989576        | removed | 2016-11-27 20:36:51 | 2016-11-27 20:36:58 |
| 101 | test610097        | removed | 2016-11-27 20:36:59 | 2016-11-27 20:37:06 |
| 111 | test636269        | removed | 2016-11-27 20:40:02 | 2016-11-27 20:40:10 |
| 112 | test794208        | removed | 2016-11-27 20:40:11 | 2016-11-27 20:40:18 |
+-----+-------------------+---------+---------------------+---------------------+
10 rows in set (0.00 sec)

mysql> select id,name,state,created,removed from service where state="removed";    
+-----+---------------+---------+---------------------+---------------------+
| id  | name          | state   | created             | removed             |
+-----+---------------+---------+---------------------+---------------------+
| 180 | LB-test318724 | removed | 2016-11-27 20:27:45 | 2016-11-27 20:28:34 |
| 188 | LB-test770316 | removed | 2016-11-27 20:28:55 | 2016-11-27 20:30:03 |
| 214 | LB-test232087 | removed | 2016-11-27 20:31:37 | 2016-11-27 20:32:50 |
| 218 | LB-test232087 | removed | 2016-11-27 20:31:55 | 2016-11-27 20:32:50 |
| 221 | LB-test827497 | removed | 2016-11-27 20:33:00 | 2016-11-27 20:33:22 |
| 224 | LB-test827497 | removed | 2016-11-27 20:33:10 | 2016-11-27 20:33:22 |
| 240 | test4         | removed | 2016-11-27 20:36:52 | 2016-11-27 20:36:59 |
| 241 | test5         | removed | 2016-11-27 20:37:00 | 2016-11-27 20:37:07 |
| 253 | test4         | removed | 2016-11-27 20:40:03 | 2016-11-27 20:40:10 |
| 254 | test5         | removed | 2016-11-27 20:40:12 | 2016-11-27 20:40:19 |
+-----+---------------+---------+---------------------+---------------------+
10 rows in set (0.00 sec)

mysql> select id,name,state,created,removed from instance where state="purged";    
+-----+-----------------------------------+--------+---------------------+---------------------+
| id  | name                              | state  | created             | removed             |
+-----+-----------------------------------+--------+---------------------+---------------------+
|   1 | scheduler-scheduler-1             | purged | 2016-11-27 18:59:54 | 2016-11-27 20:56:38 |
| 285 | native-test-799557                | purged | 2016-11-27 20:21:50 | 2016-11-27 20:22:09 |
| 338 | test951525rancher-LB-test318724-1 | purged | 2016-11-27 20:27:47 | 2016-11-27 20:28:37 |
| 348 | test724411rancher-LB-test770316-1 | purged | 2016-11-27 20:28:55 | 2016-11-27 20:30:05 |
| 367 | test471745-LB-test232087-1        | purged | 2016-11-27 20:31:38 | 2016-11-27 20:32:53 |
| 380 | test471745rancher-LB-test232087-1 | purged | 2016-11-27 20:31:56 | 2016-11-27 20:32:54 |
| 382 | test127932-LB-test827497-1        | purged | 2016-11-27 20:33:01 | 2016-11-27 20:33:25 |
| 388 | test127932rancher-LB-test827497-1 | purged | 2016-11-27 20:33:11 | 2016-11-27 20:33:26 |
| 430 | test989576-test4-1                | purged | 2016-11-27 20:36:53 | 2016-11-27 20:37:00 |
| 431 | test989576-test4-2                | purged | 2016-11-27 20:36:53 | 2016-11-27 20:37:00 |
| 432 | test610097-test5-1                | purged | 2016-11-27 20:37:00 | 2016-11-27 20:37:08 |
| 433 | test610097-test5-2                | purged | 2016-11-27 20:37:01 | 2016-11-27 20:37:08 |
| 472 | test636269-test4-1                | purged | 2016-11-27 20:40:04 | 2016-11-27 20:40:11 |
| 473 | test636269-test4-2                | purged | 2016-11-27 20:40:04 | 2016-11-27 20:40:11 |
| 474 | test794208-test5-1                | purged | 2016-11-27 20:40:13 | 2016-11-27 20:40:20 |
| 475 | test794208-test5-2                | purged | 2016-11-27 20:40:13 | 2016-11-27 20:40:19 |
+-----+-----------------------------------+--------+---------------------+---------------------+
16 rows in set (0.00 sec)

mysql> 
```

**Comment by cjellick** [comment]
TL;DR - cleanup is working as intended and as best it can with how our instance/volume life cycles work right now.
too long version:
the cleanup cleaned up over 300 instances. the 16 that were left behind were done so because they had mounts that were inactive instead of removed. a mount is the relationship between an instance and volume. to facilitate proper cleanup, we dont remove the mount until the VOLUME is removed. generally, we remove local volumes that arent in use elsewhere when the instance is purged, but with one big exception: if the volume  is a host bind mount (ie `/var:/host/var`) then we dont remove it when the instance is purged. this means the mount stays in inactive and cannot be cleaned up by the cleanup logic.

that logic can be changed or improved, but in my opinion it is out of scope for 1.2

the services and environments that were not removed were the ones tied to these instances that have host bind mounts because the cleanup works in a cascading fashion.

**Comment by cjellick** [comment]
closing per @deniseschannon

**Comment by kkvilekval** [comment]
I have been seeing quite a few of these the log which might be related.
```
2016-12-13 13:26:14,466 ERROR [:] [] [] [] [TaskScheduler-1] [i.c.p.core.cleanup.TableCleanup     ] SQL [delete from `instance` where `instance`.`id` in (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)]; Cannot delete or update a parent row: a foreign key constraint fails (`cattle`.`network_service_provider_instance_map`, CONSTRAINT `fk_network_service_provider_instance_map__instance_id` FOREIGN KEY (`instance_id`) REFERENCES `instance` (`id`) ON DELETE NO ACTIO) 
2016-12-13 13:26:14,570 WARN  [:] [] [] [] [TaskScheduler-1] [i.c.p.core.cleanup.TableCleanup     ] [Rows Skipped] host=33 instance=7 physical_host=33 service_index=100 service=100 environment=58 storage_pool=33 agent=100 account=100  
2016-12-13 14:26:14,442 ERROR [:] [] [] [] [TaskScheduler-2] [i.c.p.core.cleanup.TableCleanup     ] SQL [delete from `instance` where `instance`.`id` in (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)]; Cannot delete or update a parent row: a foreign key constraint fails (`cattle`.`network_service_provider_instance_map`, CONSTRAINT `fk_network_service_provider_instance_map__instance_id` FOREIGN KEY (`instance_id`) REFERENCES `instance` (`id`) ON DELETE NO ACTIO) 
2016-12-13 14:26:14,523 WARN  [:] [] [] [] [TaskScheduler-2] [i.c.p.core.cleanup.TableCleanup     ] [Rows Skipped] host=33 instance=7 physical_host=33 service_index=100 service=100 environment=58 storage_pool=33 agent=100 account=100  
2016-12-13 15:26:14,434 ERROR [:] [] [] [] [TaskScheduler-2] [i.c.p.core.cleanup.TableCleanup     ] SQL [delete from `instance` where `instance`.`id` in (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)]; Cannot delete or update a parent row: a foreign key constraint fails (`cattle`.`network_service_provider_instance_map`, CONSTRAINT `fk_network_service_provider_instance_map__instance_id` FOREIGN KEY (`instance_id`) REFERENCES `instance` (`id`) ON DELETE NO ACTIO) 
2016-12-13 15:26:14,514 WARN  [:] [] [] [] [TaskScheduler-2] [i.c.p.core.cleanup.TableCleanup     ] [Rows Skipped] host=33 instance=7 physical_host=33 service_index=100 service=100 environment=58 storage_pool=33 agent=100 account=100  
2016-12-13 16:26:14,442 ERROR [:] [] [] [] [TaskScheduler-2] [i.c.p.core.cleanup.TableCleanup     ] SQL [delete from `instance` where `instance`.`id` in (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)]; Cannot delete or update a parent row: a foreign key constraint fails (`cattle`.`network_service_provider_instance_map`, CONSTRAINT `fk_network_service_provider_instance_map__instance_id` FOREIGN KEY (`instance_id`) REFERENCES `instance` (`id`) ON DELETE NO ACTIO) 
```
